documentation/docs/03-routing.md

```diff
@@ -45,206 +57,170 @@ A file called either `src/routes/about.svelte` or `src/routes/about/index.svelte
+export function GET({ session }) {
+	if (session.user?.name !== 'Dan') {
+		throw error(403, `If your name's not Dan, you're not coming in`);
+	}
+
+	return {
+		answer: 42
+	};
+}
 ```
 
-These functions can, like `GET`, return a `body` that will be passed to the page as props. Whereas 4xx/5xx responses from `GET` will result in an error page rendering, similar responses to non-GET requests do not, allowing you to do things like render form validation errors:
+The first argument to `error` is an [HTTP status code](https://httpstatusdogs.com) (you'll most likely reach for `4xx` — client error, or `5xx` — server error), the second is an explanatory message. You can access both using the [page store](/docs/modules#$app-stores-page).
 
-```js
-/// file: src/routes/items.js
-// @filename: ambient.d.ts
-type Item = {
-	id: string;
-};
-type ValidationError = {};
+#### POST
 
-declare module '$lib/database' {
-	export const list: () => Promise<Item[]>;
-	export const create: (request: Request) => Promise<[Record<string, ValidationError>, Item]>;
-}
+In the case of `POST`, there are basically three possible outcomes — success, failure due to input validation errors, and failure due to something unexpected that requires showing an error page.
 
-// @filename: __types/items.d.ts
-import type { RequestHandler as GenericRequestHandler } from '@sveltejs/kit';
-export type RequestHandler<Body = any> = GenericRequestHandler<{}, Body>;
+In the success case, nothing or a redirect is returned. The latter is often used to redirect to the newly created resource. We can achieve this by returning a `location` property that indicates where the created resource lives:
 
-// @filename: index.js
-// ---cut---
-import * as db from '$lib/database';
+```js
+/// file: src/routes/todos/+page.server.js
 
-/** @type {import('./__types/items').RequestHandler} */
-export async function GET() {
-	const items = await db.list();
+/** @type {import('./$types').POST} */
+export function POST({ request }) {
+  const data = await request.formData();
+  await create_todo(data);
+  const id = await create_todo(data);
 
-	return {
-		body: { items }
-	};
+  return {
+    location: `/todos/${id}`
+  };
 }
+```
 
-/** @type {import('./__types/items').RequestHandler} */
-export async function POST({ request }) {
-	const [errors, item] = await db.create(request);
-
-	if (errors) {
-		// return validation errors
-		return {
-			status: 400,
-			body: { errors }
-		};
-	}
+After the `POST` handler returns, the `GET` handler (on the current site or the one that is redirected to) is invoked again, being able to provide the new data.
 
-	// redirect to the newly created item
-	return {
-		status: 303,
-		headers: {
-			location: `/items/${item.id}`
-		}
-	};
+Unexpected failures are easy to deal with, as it's the same as `GET` — `throw error(status)`.
+
+Validation errors cause the page to be re-rendered with the errors, so they can be used for UI that guides the user towards valid inputs:
```

Comment:
In the case of an invalid post request, how does one show the invalid value in the input? E.g., I'd want to do this...

```
<p>Error: Must include the word "potato".</p>
<input name="description" value="tomato" />
```
According to my reading of this section and the [migration guide discussion](https://github.com/sveltejs/kit/discussions/5774#:~:text=POST%2C%20PUT%20and%20PATCH%20handlers%20should%20not%20return%20data.%20They%20can%20return%20an%20errors%20object%20of%20validation%20errors), the POST handler is only allowed to return a map of validation errors, not the invalid data itself. The data is populated from GET, not passed back from what the POST handler originally read. So, it appears the user could be shown the error, but not the invalid value that they submitted.

Reply:
The plan is basically to take a different tack altogether: https://github.com/sveltejs/kit/discussions/5875

Until then you can have an `errors` prop on your page, and it will be filled in with the validation errors:

```svelte
<script>
  export let data;
  export let errors;
</script>
```

Reply:
oh shit, wait, that doesn't answer your question at all. err...

this is nasty, but you could put the values on `event.locals` and read them in the `GET`